core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/HttpSignRequest.java

```diff
@@ -20,25 +20,21 @@
 import software.amazon.awssdk.annotations.SdkPublicApi;
 import software.amazon.awssdk.annotations.ThreadSafe;
 import software.amazon.awssdk.http.SdkHttpRequest;
-import software.amazon.awssdk.http.auth.spi.internal.DefaultHttpSignRequest;
-import software.amazon.awssdk.utils.builder.SdkBuilder;
+import software.amazon.awssdk.identity.spi.Identity;
 
 /**
- * Represents a request to be signed by {@link HttpSigner}.
+ * Input parameters to sign a request using {@link HttpSigner}.
  *
- * @param <PayloadT> The type of payload of this request.
+ * @param <PayloadT> The type of payload of the request.
+ * @param <IdentityT> The type of the identity.
  */
+// TODO: Code Reviewer Note: Seems like this could be package-private/SdkProtectedApi, since there are 2 sub-interfaces that
+//  HttpSigner uses, however for implementation of HttpSigner that don't care about Sync v/s Async, they can share their
+//  implementation by using this type with generic PayloadT. Should this be public but SdkProtectedApi?
 @SdkPublicApi
 @Immutable
 @ThreadSafe
-public interface HttpSignRequest<PayloadT> {
-
-    /**
-     * Get a new builder for creating a {@link HttpSignRequest}.
-     */
-    static <T> Builder<T> builder(Class<T> payloadType) {
-        return new DefaultHttpSignRequest.BuilderImpl(payloadType);
-    }
+public interface HttpSignRequest<PayloadT, IdentityT extends Identity> {
```

Comment:
Imo, this should implement `ToCopyableBuilder` and its builder should implement `CopyableBuilder`. That prevents some issues here, needing to remember to override every parent method in the child class.

Unfortunately it makes the parameters on this type even crazier.

To make the room, we could change `IdentityT identity()` to `Identity identity()` and `Optional<PayloadT> payload()` to `Optional<? extends Object> payload()`, and then override them in the child types to return the exact types. Even better, we might just remove the getters from here entirely and rely on the subtype. I don't expect it to be a major problem to need to use the subtype to get access to the payload or identity.

We'd also need to remove the `payload` and `identity` setters from this base builder, but that's probably okay - we can just put them in the child types. It should be very rare that someone needs to set the identity or payload using the base builder type...

Reply:
> That prevents some issues here, like needing to remember to override every parent method in the child class.

Without using To/CopyableBuilder, I was able to remove the overriding of methods in the builder sub-interfaces to return the sub-type by adding a generic type parameter for the builder, in these commits https://github.com/aws/aws-sdk-java-v2/pull/4018/commits/de704e738dbb15530a83ee86f95c9606ccd3cb39 and https://github.com/aws/aws-sdk-java-v2/pull/4018/commits/3e24cc490fcc665f60e9a735e7d2d4a8dcff6ba9.

I'm not sure if we'll really need to mutate these objects, so I think we can defer making these To/CopyableBuilder until needed?